File: python/day5.py
# Advent of Code
# Day 5: If you give a seed a fertilizer

import re
from itertools import chain

SEEDINPUT = re.compile(r"seeds: ([\d ]*)")
MAPINPUT = re.compile(r"([\w-]+) map:")

# Why is calling `min` on a range object so slow?
# I had this mostly solved but it was taking so long
# to run I thought I needed to do something else. 

def range_overlap(left, right):
    return range(max(right.start,left.start), min(right.stop,left.stop))

class AlmanacSeeds:
    def __init__(self):
        self.ranges = set()
    def add(self, range_start, range_len):
        self.ranges.add(range(range_start, range_start+range_len))
    def __iter__(self):
        return chain.from_iterable(self.ranges)
    def has(self, item):
        for r in self.ranges:
            if item in r:
                return True
        return False

class AlmanacRange:

    ranges = {}

    def __new__(cls, dst_start, src_start, range_len):
        try:
            return cls.ranges[ (dst_start, src_start, range_len) ]
        except KeyError:
            pass
        self = super().__new__(cls)
        self.dst = range(dst_start, dst_start+range_len)
        self.src = range(src_start, src_start+range_len)
        cls.ranges[ (dst_start, src_start, range_len) ] = self
        return self

class AlmanacMap:

    def __init__(self, input):
        self.ranges = set()
        while line := input.readline().rstrip():
            self.add(*map(int, line.split()))

    def add(self, *args):
        if isinstance(args[0], AlmanacRange):
            self.ranges.add(args[0])
        else:
            self.ranges.add(AlmanacRange(*args))

    def __getitem__(self, item):
        for r in self.ranges:
            if item in r.src:
                return r.dst.start + r.src.index(item)
        return item

    def overlap(self, items):
        items = [items]
        dst = []
        while items:
            s = items.pop()
            for r in self.ranges:
                o = range_overlap(r.src, s)
                if o:
                    d = r.dst.start+r.src.index(o.start)
                    dst.append( range(d, d+len(o)) )
                    if s.start < o.start:
                        items.append( range(s.start, o.start) )
                    if o.stop < s.stop:
                        items.append( range(o.stop, s.stop) )
                    break
            else:
                dst.append(s)
        return dst

class Almanac1:

    def __init__(self, input):
        self.seeds = self.input_seeds(input)
        #print("Got seeds:",self.seeds)
        self.maps = {}
        while line := input.readline():
            m = MAPINPUT.match(line)
            if m:
                #print("Adding map:", m[1])
                self.maps[m[1]] = AlmanacMap(input)

    def input_seeds(self, input):
        while line := input.readline().rstrip():
            m = SEEDINPUT.match(line)
            if m:
                return [int(s) for s in m[1].split()]
        return []

    def lookup(self, item, table):
        src = table[0]
        for dst in table[1:]:
            #print(src,item,dst)
            m = self.maps[src+'-to-'+dst]
            item = m[item]
            src = dst
        return item

    def lookup_range(self, items, table):
        src = table[0]
        for dst in table[1:]:
            overlap = []
            m = self.maps[src+'-to-'+dst]
            for r in items:
                #print(src,r,dst)
                overlap.extend(m.overlap(r))
            items = overlap
            src = dst
        return items

class Almanac2(Almanac1):
    def input_seeds(self, input):
        seeds = AlmanacSeeds()
        while line := input.readline().rstrip():
            m = SEEDINPUT.match(line)
            if m:
                ranges = [int(s) for s in m[1].split()]
                for pair in zip(ranges[::2],ranges[1::2]):
                    seeds.add(*pair)
        return seeds

def part1(input):
    almanac = Almanac1(input)
    seed_to_location = lambda s: almanac.lookup(s, 'seed soil fertilizer water light temperature humidity location'.split())
    return min(seed_to_location(seed) for seed in almanac.seeds)

def part2(input):
    almanac = Almanac2(input)
    seed_to_location = lambda s: almanac.lookup_range(s, 'seed soil fertilizer water light temperature humidity location'.split())
    return min(r.start for r in seed_to_location(almanac.seeds.ranges))

def main():
    import sys
    stage = part1
    if len(sys.argv) > 1:
        if sys.argv[1] == '1':
            stage = part1
        elif sys.argv[1] == '2':
            stage = part2
        else:
            return
    print(stage(sys.stdin))

if __name__=='__main__':
    main()

File: lua/day5.lua
local aoc = require"adventofcode"

function input_seeds(file)
    local seeds
    line = file:read()
    while line do
        if line:match("seeds:") then
            seeds = aoc.map(aoc.findall(line, "%d+"), tonumber)
            break
        end
        line = file:read()
    end
    return seeds
end

function input_map(file)
    local map, name = {}
    line = file:read()
    while line do
        name = line:match("([%S]+) map:")
        if name then
            break
        end
        line = file:read()
    end
    if not name then
        return nil
    end
    line = file:read()
    while line do
        if line == "" then
            break
        end
        local dst, src, len = line:match("(%d+) (%d+) (%d+)")
        map[#map+1] = {dst = tonumber(dst), src = tonumber(src), n = tonumber(len)}
        line = file:read()
    end
    return name, map
end

local function split_map_range(range, len)
    if len > 0 and len < range.n then
        local left = {dst = range.dst, src = range.src, n = len}
        local right = {dst = range.dst+len, src = range.src+len, n = range.n-len}
        --print(string.format("{%d,%d,%d}->{%d+%d}/{%d+%d}",
        --      range.src,range.dst,range.n,left.src,left.n,right.src,right.n))
        return left,right
    end
    return range
end

function map_sort_dst(map)
    table.sort(map, function(i,j) return i.dst < j.dst end)
end

function map_sort_src(map)
    table.sort(map, function(i,j) return i.src < j.src end)
end

function merge_map(left, right)
    map_sort_dst(left)
    map_sort_src(right)
    local merged = {}
    local lr, next = left[1], 2
    for _,rr in ipairs(right) do
        while lr and rr.src+rr.n > lr.dst do
            -- Left side is behind or overlapping right
            if lr.dst+lr.n <= rr.src then
                -- Left range completely behind right side, catch up
                merged[#merged+1] = lr
                lr, next = left[next], next+1
            else
                -- Left and right overlap, transform them
                if lr.dst < rr.src then
                    -- Leading left
                    local sp
                    sp, lr = split_map_range(lr, rr.src-lr.dst)
                    merged[#merged+1] = sp
                end
                if rr.src < lr.dst then
                    -- Leading right
                    local sp
                    sp, rr = split_map_range(rr, lr.dst-rr.src)
                    merged[#merged+1] = sp
                end
                assert(lr.dst == rr.src)
                if lr.n < rr.n then
                    local sp
                    sp, rr = split_map_range(rr, lr.n)
                    sp.src = lr.src
                    merged[#merged+1] = sp
                    lr, next = left[next], next+1
                else
                    local sp
                    sp, lr = split_map_range(lr, rr.n)
                    sp.dst = rr.dst
                    merged[#merged+1] = sp
                    if not lr then
                        lr, next = left[next], next+1
                    end
                    rr = nil
                    break
                end
            end
        end
        if rr and (not lr or rr.src+rr.n <= lr.dst) then
            -- Left range completely in front of right side, pass through
            merged[#merged+1] = rr
        end
    end
    while lr do
        -- Remaining left side ranges
        merged[#merged+1] = lr
        lr, next = left[next], next+1
    end
    return merged
end

function lookup(map, value)
    for _,range in ipairs(map) do
        if value >= range.src and value < range.src+range.n then
            return value + range.dst - range.src
        end
    end
    return value
end

function input_almanac(file)
    local seeds = input_seeds(file)
    local maps = {}
    for name, map in input_map,file do
        maps[name] = map
    end
    local seed_map = maps['seed-to-soil']
    local map_name = 'soil'
    for _,name in ipairs{'fertilizer'
    ,'water'
    ,'light'
    ,'temperature'
    ,'humidity'
    ,'location'
    } do
        --print(map_name,name)
        seed_map = merge_map(seed_map, maps[map_name..'-to-'..name])
        map_name = name
    end
    return seeds, seed_map
end

function part1(seeds, seed_map)
    local location = math.huge
    for _,s in ipairs(seeds) do
        local l = lookup(seed_map, s)
        if l < location then location = l end
    end
    return location
end

local function overlap(left,right)
    return left[1] < right[2] and right[1] < left[2]
end

function part2(seeds, seed_map)
    local seed_ranges = {}
    for i = 1,#seeds-1,2 do
        seed_ranges[#seed_ranges+1] = {seeds[i], seeds[i]+seeds[i+1]}
    end
    map_sort_dst(seed_map)
    for _,loc in ipairs(seed_map) do
        local l = {loc.src, loc.src+loc.n}
        for _,r in ipairs(seed_ranges) do
            if overlap(l, r) then
                if l[1] > r[1] then
                    return loc.dst
                else
                    return loc.dst + r[1] - l[1]
                end
            end
        end
    end
end

-- main
local stage = part1
if arg[1] == '2' then stage = part2 end
local seeds, seed_map = input_almanac(io.stdin)
print(stage(seeds, seed_map))

